Import MultiheadAttention to PyTorch

zhangguanheng66 opened the pull request:
Summary:
Import MultiheadAttention into the core pytorch framework.
Users now can import MultiheadAttention directly from torch.nn.
See "Attention Is All You Need" for more details related to MultiheadAttention function.

Differential Revision: D14577966

soumith commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +669,269 @@ def extra_repr(self):
+    See "Attention Is All You Need" for more details.
+    """
+
+    def __init__(self, embed_dim, num_heads, dropout=0., bias=True, add_bias_kv=False, add_zero_attn=False):
```
constructor needs documentation

zhangguanheng66 replied:
Add constructor for MultiheadAttention.

soumith commented on torch/nn/functional.py:
```diff
@@ -1408,6 +1408,22 @@ def linear(input, weight, bias=None):
 
 
+@weak_script
+def linear_weight_float(input, weight, bias=None):
```
if this is only for MultiHeadAttention, prefix with `_` and make it internal. Right now, the API and semantics of this function dont look great.

zhangguanheng66 replied:
The function is not needed any more.

soumith commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +669,269 @@ def extra_repr(self):
+
+    def forward(self, query, key, value, key_padding_mask=None, incremental_state=None,
+                need_weights=True, static_kv=False, attn_mask=None):
+        """Input shape: Time x Batch x Channel
```
needs better docs

zhangguanheng66 replied:
See the description/document for the construction.

soumith commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +669,269 @@ def extra_repr(self):
+
+        if attn_mask is not None:
+            attn_mask = attn_mask.unsqueeze(0)
+            if self.onnx_trace:
```
i think self.onnx_trace should probably be removed?

zhangguanheng66 replied:
Since reorder_incremental_state() function is not tested in the unit test, I remove it from the class. Therefore, self.onnx_trace is removed as well.

soumith reviewed (changes requested):
looks like a WIP PR. Next time, prefix the title with [WIP] so that reviewers dont end up reviewing prematurely :)

zhangguanheng66 commented:
Thanks @soumith . Will accommodate your review soon.

zhangguanheng66 commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +672,269 @@ def extra_repr(self):
+    """
+
+    def __init__(self, embed_dim, num_heads, dropout=0., bias=True, add_bias_kv=False, add_zero_attn=False):
+        super().__init__()
```
super(MultiheadAttention, self).__init__()

cpuhrsch commented:
@zhangguanheng66 - just because it deserves explicit mention: since this is a straight-up port from fairseq, before making any major changes to the code, make sure you wrote a ton of tests so we don't get lost.

cpuhrsch commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +673,284 @@ def extra_repr(self):
+            bias = bias[start:end]
+        return F.linear(input, weight, bias)
+
+    def reorder_incremental_state(self, incremental_state, new_order):
```
This function isn't covered by unit tests and exercises some internal methods we might not need to ship right away as well. We can bring this back later on, once we need it.

cpuhrsch replied:
This might apply to various other methods as well. If the unit test doesn't exercise a method, remove it and mark it for something we might want to add later on via documentation (`# TODO: <COMMENT>`)

zhangguanheng66 replied:
Any functions which are not tested have been removed from MultiheadAttention.

zhangguanheng66 commented:
@cpuhrsch @soumith Thanks for the feedbacks. I updated the code accordingly. Functions that are not tested have been removed. A unit test is created in test.nn. An additional test was conducted on language_technology/neural_mt/os/pytorch_translate. Instead of using fairseq.MultiheadAttention, torch.nn.MultiheadAttention is applied and the corresponding unit test in pytorch_translate works fine.

cpuhrsch commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +673,210 @@ def extra_repr(self):
+
+        return attn, attn_weights
+
+    def in_proj_qkv(self, query):
```
there's 'in_proj_*' aren't exercised in the tests either right?

So I guess we should do another pass where we also make functions private if they aren't tested, i.e. prefix them with an '_'.

zhangguanheng66 replied:
Made some changes accordingly.

cpuhrsch reviewed (approved):
I'm approving this under the assumption that the most recent comments will also be resolved.

Pinging @soumith in case something is still missing.

soumith commented:
if you dont mind, before this lands I'd like to page some folks in the NLP community to be assured that they dont need any more features from this.

cc: @srush @kyunghyuncho @myleott @glample

soumith reviewed (changes requested):
the documentation for `forward` is still missing, and `forward` takes a lot of options. Please flesh out.

zhangguanheng66 commented:
@soumith Sure. Let me know if any new features are necessary. I will work on the documentation for forward function at the same time.

@srush @kyunghyuncho @myleott @glample More unit tests on your end are welcome (I have two unit tests on my side).

kyunghyuncho commented:
@mansimov @jasonleeinf care to take a quick look at this PR?

myleott commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +673,210 @@ def extra_repr(self):
+            attn_weights = attn_weights.float().masked_fill(
+                key_padding_mask.unsqueeze(1).unsqueeze(2),
+                float('-inf'),
+            ).type_as(attn_weights)  # FP16 support: cast to float and back
```
I think this hack may not be needed anymore:
```
>>> torch.cuda.HalfTensor(5).masked_fill_(torch.cuda.ByteTensor([0, 1, 0, 1, 0]), float('-inf'))
tensor([0., -inf, 0., -inf, 0.], device='cuda:0', dtype=torch.float16)
```

Previously this didn't work because masked_fill_ would complain about overflow when writing -inf values, so we cast to float and back.

zhangguanheng66 replied:
Hi @myleott, so I could do like this ?

attn_weights = attn_weights.float().masked_fill(
                key_padding_mask.unsqueeze(1).unsqueeze(2),
                `'-inf',`
            ).type_as(attn_weights)

cpuhrsch replied:
We can also earmark this for a future PR, because it appears performance related, but not correctness. We want to make sure we ship something correct first.

myleott commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +673,210 @@ def extra_repr(self):
+        if attn_mask is not None:
+            attn_mask = attn_mask.unsqueeze(0)
+            # if self.onnx_trace:
+            #     attn_mask = attn_mask.repeat(attn_weights.size(0), 1, 1)
```
remove

zhangguanheng66 replied:
Good catch. I have deleted that line.

myleott commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +673,210 @@ def extra_repr(self):
+            ).type_as(attn_weights)  # FP16 support: cast to float and back
+            attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
+
+        attn_weights = F.softmax(attn_weights.float(), dim=-1).type_as(attn_weights)
```
The native FP32 softmax is preferred now:
```F.softmax(attn_weights, dim=-1, dtype=torch.float32).type_as(attn_weights)```

soumith replied:
dtype=torch.float32 only if attn_weights.dtype != torch.float64

myleott replied:
Ah, good call. Or is it better to check for float16 explicitly?

Or maybe leave all this to apex? I'm still holding out for an apex.convert_ops_to_fp32 context manager...

soumith replied:
i think the dtype line can safely be:

```
dtype=torch.float32 if attn_weights.dtype == torch.float16 else torch.float32
```
```

zhangguanheng66 replied:
Thanks Soumith. I add it.

ngimel replied:
In fact, even this 
```
F.softmax(attn_weights, dim=-1, dtype=torch.float32).type_as(attn_weights)
``` 
is not needed. What's happening with this line is softmax is computed in fp32, output in fp32 and immediately truncated to fp16 if attn_weights were fp16, no-op otherwise (I'm ignoring double here on purpose, there should be no type casts in double). 
```
F.softmax(attn_weights, dim=-1)
```
would do exact same thing. `dtype=torch.float32` output type for logsoftmax/softmax should be used only if you intend to do something with fp32 output (e.g. compute loss), not if you intend to cast it back to narrower type right away.

myleott reviewed (changes requested):
I didn't review the logic carefully, but let's test this in fairseq. Ideally we'll replace the multihead attention implementation in fairseq with this one.

Evpok commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +673,210 @@ def extra_repr(self):
+
+        self._reset_parameters()
+
+    def _reset_parameters(self):
```
Why use `xavier_uniform` here when [`Linear`](https://pytorch.org/docs/stable/_modules/torch/nn/modules/linear.html#Linear)'s default is `kaiming_uniform`?

myleott replied:
This is what they use in the original tensor2tensor implementation, which is what we have used in fairseq as well.

myleott replied:
This is what they use in the original tensor2tensor implementation, which is what we have used in fairseq as well.

ngimel replied:
except for initialization quirk based on parameter matrix size, that I've mentioned earlier.

zhangguanheng66 commented:
@myleott Thanks for reviewing. I will make the changes accordingly. Although I have tested the new module on a translate application, please let me know if you have more cases to tests. I don't see a lot of applications coming with the corresponding unit tests.

ngimel commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +673,225 @@ def extra_repr(self):
+        assert self.head_dim * num_heads == self.embed_dim, "embed_dim must be divisible by num_heads"
+        self.scaling = self.head_dim ** -0.5
+
+        self.in_proj_weight = Parameter(torch.Tensor(3 * embed_dim, embed_dim))
```
Fairseq softmax employs this perf optimization that is carried over here, where 3 matrix multiplies are combined into 1, with a bigger weight (similar to how this is done e.g. for LSTM). Mathematically, the operations performed are exactly the same, however, initialization could be different if it is based on the first dimension of the parameter matrix. IIRC, tf t2t transformer has 3 weight matrices here instead of one, with correspondingly different initialization. I'm not asking to change anything for this PR, I just want us to be aware of cross-framework initialization differences, and of potential changes that might happen here if we decide to go to 3 matrices instead of 1 (believe it or not, there can be perf reasons to go back to 3 matrix multiplies).

cpuhrsch replied:
This might be important to decide at this point right? Presumably this will be hard to change in the future, unless we create an option to switch between the two.

ngimel replied:
If we want core multihead attention to replicate the behavior of fairseq, then our hands are tied already. I'd defer to @myleott for a question of what initialization behavior we want, and what should be the options.

myleott replied:
@ngimel makes a great point. We can decouple these, by initializing the weights of each portion separately.

Lemme run a quick test to see how this affects NMT model quality. I’ll share findings back here ~tomorrow.

zhangguanheng66 replied:
Hi @myleott. I just want to reach out and see if the decision has been made.

myleott replied:
It actually does seem to make a difference, with the fairseq version producing better results by ~0.9 BLEU on WMT'16 En-De. It's possible that this could be corrected by adjusting the learning rate, although I didn't try this yet.

It's up to you how to proceed. We'll have fairseq switch to this implementation in any case, and use a custom weight init if needed.

zhangguanheng66 replied:
Thanks for the update @myleott . I will change to three individual matrices for the weights and initiate them separately.

myleott replied:
I'm not sure you should actually make it three separate matrices, since that will be slower. You can instead initialize them as if there were separate though:
```
nn.init.xavier_uniform_(self.in_proj_weight[:embed_dim, :])
nn.init.xavier_uniform_(self.in_proj_weight[embed_dim:2*embed_dim, :])
nn.init.xavier_uniform_(self.in_proj_weight[2*embed_dim:, :])
```

myleott replied:
I looked into this a bit more and it's actually a much smaller difference, only ~0.3 BLEU worse using the split init vs. combined init.

zhangguanheng66 replied:
I will just keep it as the split init. Thanks for investigating this.

zhangguanheng66 commented:
> the documentation for `forward` is still missing, and `forward` takes a lot of options. Please flesh out.

I add some documentation under forward function. Thanks. @soumith

zou3519 commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +673,225 @@ def extra_repr(self):
+                key_padding_mask = torch.cat(
+                    [key_padding_mask, torch.zeros(key_padding_mask.size(0), 1).type_as(key_padding_mask)], dim=1)
+
+        attn_weights = torch.bmm(q, k.transpose(1, 2))
```
Instead of using `bmm`, you could use `matmul`. This would make some of the views above nicer (right now the batch / head dimensions get flattened so that they can be sent into `bmm`, but with `matmul` there is no need to flatten.

aszlam commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +673,225 @@ def extra_repr(self):
+            static_kv: key and value are static
+
+        Outputs of forward function
+            attn: [embed_dim, sequence length, head_dim]
```
is this right?  why is the attn output a different format than the input?  Also, why is the name of the output "attn"?

zhangguanheng66 replied:
There must be some typos there. I will take a look at it.

myleott replied:
Also, I think we call it attn in fairseq, but let's rename it to something friendly (output?).

zhangguanheng66 replied:
The documentation has been modified. 
Use attn_output and attn_output_weights.

fmassa commented:
Also, could you use `torch.empty` or `torch.tensor` instead of the legacy `torch.Tensor` for creating a tensor?

zhangguanheng66 commented:
> Also, could you use `torch.empty` or `torch.tensor` instead of the legacy `torch.Tensor` for creating a tensor?

zhangguanheng66 commented:
@soumith @myleott Please let me know if you have additional suggestions before landing this PR.

zou3519 commented on torch/nn/modules/activation.py:
```diff
@@ -669,6 +672,231 @@ def extra_repr(self):
+            attn_output_weights = attn_output_weights.float().masked_fill(
+                key_padding_mask.unsqueeze(1).unsqueeze(2),
+                float('-inf'),
+            ).type_as(attn_output_weights)  # FP16 support: cast to float and back
```
This seems inefficient if we're not doing fp16. Can we add conditionals against the data type?

zhangguanheng66 replied:
Since the masked_fill doesn't complain about overflow, I don't need to convert the weights to float and then back.

myleott replied:
Yep, this used to be a problem where you couldn't write -inf to HalfTensors, but it's since been fixed, fine to remove.

zhangguanheng66 committed:
Import MultiheadAttention to PyTorch (#18334)

Summary:
Import MultiheadAttention into the core pytorch framework.
Users now can import MultiheadAttention directly from torch.nn.
See "Attention Is All You Need" for more details related to MultiheadAttention function.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/18334

Differential Revision: D14577966

fbshipit-source-id: b18d945ea461c07948d2f33f5b497ca51591d0ce

vadimkantorov commented:
The doc string is there, but it seems missing the `rst` `automodule` entry @soumith

tshrjn commented:
Missing documentation for `attn_mask`, `add_zero_attn` and `add_bias_kv`.

zhangguanheng66 commented:
> Missing documentation for `attn_mask`, `add_zero_attn` and `add_bias_kv`.

We had a PR to update the attn_mask.

deepwilson commented:
What is the intuition behind adding "MultiheadAttention" block under activation.py? @zhangguanheng66
